Documentation for the new interactive UI elements (#33)
aseriousbiz/abbot-docs: src/guides/ui-elements/index.md

---
uid: ui-elements
---

# User Interface Elements

> Abbot supports a limited set of UI elements that can be leveraged in a skill. Unfortunately, these elements only work for Slack and Teams. Discord doesn't yet support UI elements. We may have a workaround for that in the future, but for now, this is the way it is.

Reviewer: > Unfortunately, these elements only work for Slack and Teams

This feels more factual than unfortunate (it's not the subject of fortune at all). This could be restated as "Discord doesn't support UI elements yet, so these elements only work in Slack and Teams."

> Over time, Abbot will add support for more and more elements.

Reviewer: ```suggestion
Over time, Abbot will add support for more elements.
```

## Buttons

Buttons are a simple way to present the user a set of choices. They can be useful in workflows driven by chat such as an approval workflow. For example, suppose a skill needs to find out the user's favorite color. The following code will reply with some text and a set of buttons.

# [C#](#tab/tabid-cs)

```cs
var buttons = new[] {
    new Button("Red"),
    new Button("Blue"),
    new Button("Green")
};
await Bot.ReplyWithButtonsAsync("Please choose a color.", buttons);
```

# [JavaScript](#tab/tabid-js)

```javascript

> module.exports = (async () => { // We modularize your code in order to run it.

Reviewer: If the user copies and pastes this example directly into a new JavaScript skill, I think they will get errors. We add the first and last line in the editor to make it clearer how we run their code, and so that jslint doesn't freak out when they try to do async calls outside of a function.

Not sure if we want to include those lines in the examples or not.

Author: Yeah, we should omit them. Especially if we come up with a solution down the road where we don't need to show those lines to the user in the editor.

When the user clicks a button, the same skill is called with the value of the button. There's also a new `Bot.IsInteraction` (C#), `bot.is_interaction` (Python), `bot.isInteraction` (JavaScript) property that indicates whether or not the incoming message is the result of clicking on a button. Let's put these together.

# [C#](#tab/tabid-cs)

```cs
if (!Bot.IsInteraction) {
    var buttons = [
        new Button("Red"),
        new Button("Blue"),
        new Button("Green")
    ];
    await Bot.ReplyWithButtonsAsync("Please choose a color.", buttons);
}
else {
    // Bot.IsInteraction is true
    await Bot.ReplyAsync($"You chose {Bot.Arguments[0]}");
}
```

# [JavaScript](#tab/tabid-js)

```javascript
module.exports = (async () => { // We modularize your code in order to run it.
    if (!bot.isInteraction) {
        var buttons = [
            new Button("Red"),
            new Button("Blue"),
            new Button("Green")
        ];
        bot.replyWithButtons("Please choose a color.", buttons);
    }
    else {

>         // Bot.IsInteraction is true

Reviewer: ```suggestion
        // bot.isInteraction is true
```

        bot.reply(`You chose ${bot.tokenizedArguments[0]}`);
    }
})(); // Thanks for building with Abbot!
```

# [Python](#tab/tabid-py)

```python
if not bot.is_interaction:
    buttons = [Button("Red"), Button("Blue"), Button("Green")]
    bot.reply_with_buttons("Please choose a color", buttons)
else:

>     # bot.is_interaction false

Reviewer: ```suggestion
    # bot.is_interaction is False
```

    bot.reply("You chose {}".format(bot.tokenized_arguments[0]))
```

***

Now when you run this skill, and then click on the button, say the Blue one, you'll receive a response that says "You chose Blue".

> The value of a button is passed back to the skill as the arguments. The arguments for a button can be different from the label displayed to the user. And buttons can be given one of three styles, default, primary, or danger. However, these styles only work for Slack as Teams doesn't support them and just ignores these styles.

Reviewer: Perhaps add a sentence to clarify what "passed back to the skill as the arguments" means -- it may not be obvious to the user that a button click is similar to calling the skill with some predefined arguments.

Author: Good call. I added a bit more to really spell it out.

# [C#](#tab/tabid-cs)

```cs
if (!Bot.IsInteraction) {
    var buttons = new[] {
        new Button("Red", "pick red", ButtonStyle.Primary),
        new Button("Blue", "pick blue"),
        new Button("Green", "pick green")
    };
    await Bot.ReplyWithButtonsAsync("Please choose a color.", buttons);
}
else {
    // Bot.IsInteraction is true
    await Bot.ReplyAsync($"You chose {Bot.Arguments[1]}");
}
```

# [JavaScript](#tab/tabid-js)

```javascript
module.exports = (async () => { // We modularize your code in order to run it.
    if (!bot.isInteraction) {
        var buttons = [
            new Button("Red", "pick red", 'primary'),
            new Button("Blue", "pick blue"),
            new Button("Green", "pick green")
        ];
        bot.replyWithButtons("Please choose a color.", buttons);
    }
    else {

>         // Bot.IsInteraction is true

Reviewer: ```suggestion
        // bot.isInteraction is true
```

        bot.reply(`You chose ${bot.tokenizedArguments[1]}`);
    }
})(); // Thanks for building with Abbot!
```

# [Python](#tab/tabid-py)

```python
if not bot.is_interaction:
    buttons = [
      Button("Red", "pick red", "primary"),
      Button("Blue", "pick blue"),
      Button("Green", "pick green")
    ]
    bot.reply_with_buttons("Please choose a color", buttons)
else:

>     # bot.is_interaction false

Reviewer: ```suggestion
    # bot.is_interaction is False
```

    bot.reply("You chose {}".format(bot.tokenized_arguments[1]))
```

***

> Note that in this example, because the value has two words, I needed to grab the second tokenized argument.

Reviewer: ```suggestion
Note that in this example, because the value has two words, we need to grab the second tokenized argument.
```

# [C#](#tab/tabid-cs)

```cs
if (!Bot.IsInteraction) {
    var buttons = new[] {
        new Button("Red", "pick red", ButtonStyle.Primary),
        new Button("Blue", "pick blue"),
        new Button("Green", "pick green")
    };
    await Bot.ReplyWithButtonsAsync(
        "Please choose a color.",
        buttons,
        "Pick one",
        new Uri("https://ab.bot/img/abbot-full-wave.png"),
        "The big choice",
        "#660000");
}
else {
    // Bot.IsInteraction is true
    await Bot.ReplyAsync($"You chose {Bot.Arguments[1]}");
}
```

# [JavaScript](#tab/tabid-js)

```javascript
module.exports = (async () => { // We modularize your code in order to run it.
    if (!bot.isInteraction) {
        var buttons = [
            new Button("Red", "pick red", 'primary'),
            new Button("Blue", "pick blue"),
            new Button("Green", "pick green")
        ];
        bot.replyWithButtons("Please choose a color.",
            buttons,
            "Pick one",
            "https://ab.bot/img/abbot-full-wave.png",
            "The big choice",
            "#660000");
    }
    else {

>         // Bot.IsInteraction is true

Reviewer: ```suggestion
        // bot.isInteraction is true
```

        bot.reply(`You chose ${bot.tokenizedArguments[1]}`);
    }
})(); // Thanks for building with Abbot!
```

# [Python](#tab/tabid-py)

```python
if not bot.is_interaction:
    buttons = [
      Button("Red", "pick red", "primary"),
      Button("Blue", "pick blue"),
      Button("Green", "pick green")
    ]
    bot.reply_with_buttons("Please choose a color",
        buttons,
        "Pick one",
        "https://ab.bot/img/abbot-full-wave.png",
        "The big choice",
        "#660000")
else:

>     # bot.is_interaction false

Reviewer: ```suggestion
    # bot.is_interaction is False
```

    bot.reply("You chose {}".format(bot.tokenized_arguments[1]))
```

> Note that the final color parameter is slack only. It's ignored by Teams.

Reviewer: ```suggestion
Note that the final color parameter is Slack only. It's ignored by Teams.
```
